### ISSUE TITLE
[Proposal] Opportunistic automatic key verification by SMS

### ISSUE THREAD

**Issue by rkunze** [deltachat/deltachat-core#77]
# Opportunistic automatic key verification by SMS
The typical Deltachat installation scenario (Deltachat is installed on a smartphone) offers a ready-made out-of-band channel: SMS. This proposal outlines how SMS can be used to implement automatic key verification in Deltachat.

## Usage scenario and prerequisites
Alice and Bob want to establish a secure chat. Both have Deltachat installed on their smart phones. Both (or rather, their Deltachat messengers - from now on, I'll use "Alice" and "Bob" as shorthand for "Alices/Bobs Deltachat messenger") have already exchanged public keys. They have not yet verified that they have the correct public keys (I'll call this state an "unverified key" from now on).

In addition, Bobs contact information for Alice includes a mobile phone number.

## Key verification protocol
When Bob receives a (signed and encrypted) message from Alice that is signed by an unverified key, and Bobs contact information associated with Alices email address contains a mobile phone number, Bob initiates the key verification process by sending a SMS with a _key verification message_ to this mobile phone number. 

The key verification message consists of three parts:
1. A header containing Bobs email address and some explanatory text (e.g. "Deltachat key verification from bob@example.org")
2. A random number (of sufficient size) chosen by Bob
3. Bobs signature of both the cleartext and header and the encrypted random number

The random number and the signature are combined into a single "secret" block, encrypted with Alices public key, and encoded in a format suitable for transport over SMS (e.g. base64, or hex). The first part of the message stays in clear text. Example:
```
Deltachat key verification from bob@example.org

lJiCyqSfgtPTfmW/74I9PAAn3OyTC/Q4ZR6yMgOqqgnKtTVuH1JVV4IGPFsa9Ln+CfnMrIqJKAdg
Jk4HMBFYPjwSqzCMAZypA4xnKmCm/Dega4lLnLI4vitu+r2IqNWfhkFCpcI+M6P6Q1UxnTERAjqe
vo0sndlviciuMvH0OM7bVQSJAW0M7V/2p6iXmdShyYOuYhUYJ4j9wKSPjsQ6Ta6rP+1FgRIb2xs6
nRtFdhadLuMQSG4p8cHi66KHcjulRcDw18hJpBc31ylyB/XLhAJKOZrI7Ts3YqARSvmt3UCmbT29
pghNBHnqR/SBDKsEFzDrjaQsrFzCy3Yb9T9PoCBB4g1vl96yYd/pTunH5tosabEazME+hxjUKSp/
```

When Alice receives a key verification SMS, she checks that all of the following conditions hold:
1. The email in the cleartext header belongs to an existing chat contact (i.e. Alice already has a public key for this email address - Bobs in our example)
2. Alice can sucessfully decrypt the encrypted part
3. Alice can verify the signature with the public key she has for the email address from the message

If any of the verification conditions are _not_ met, Alice simply ignores the key validation SMS and takes no further action.

When all three conditions are met, Alice  builds a _key verification response_ and sends it back to Bob by email. The key verification response consists of the random number Alice received from Bob, is signed by Alice, and encrypted using Bobs private key.

Note: Alice does _not_ mark Bobs public key as "validated" in this step. See [here](#mallory-the-malicious-email-provider) for a discussion why.

When Bob receives a key verification response from Alice, he checks that all of the following conditions hold:
1. the message can be decrypted
2. the signature is valid against Alices public key
3. the random number is the same as Bob sent out in the key verification SMS

If all three conditions hold, Bob marks Alices key as "valid". Otherwise, Bob leaves the status for Alices key at "unverified".

Note: This protocol can only validate keys, but it can _not_ distinguish between a failure because of some technical problem (e.g. because Bob had the wrong mobile phone number for Alice, or the key verification response was lost in transit) and a man-in-the-middle attack. It may be tempting to interpret the case of "key verification SMS received, but signature verification failed" as an indicator for a man-in-the-middle attack and react accordingly by invalidating the compromised keys, but this would [allow an easy DoS attack](#dave-the-disruptor-of-service).

## Threat model
This is by no means a full threat model, just a short discussion of the impact of opportunistic key verification by SMS in some exemplary scenarios.

### Alice the intended communication partner
Alice is the intended communication partner, and as such not really an adversary. However, compared to unverified Deltachat (i.e. plain Autocrypt), Alice gains one piece of information that she might not already have: Bobs mobile number (because it is the sender of the key verification SMS). 

### Eve the passive email eavesdropper
For Eve, there are three scenarios:
1. Eve can read all email communication, but can neither modify emails nor read SMS. In this case, Eve gains no additional information compared to plain Autocrypt, because the key verification response (i.e. the only part of the process that Eve can access) is encrypted with Bobs public key, and Eve does not have access to Bobs private key.
2. Eve can read all SMS communication, but has no access to emails (this covers the case of "accidental eavesdropping on SMS", e.g. because Bob sent the key verification SMS to the wrong mobile number). In this case, Eve can gains both Bobs email address (from the cleartext SMS header) and Bobs mobile phone number (as the SMS sender), and can guess at a correlation between the two.
3. Eve can listen to both email and SMS communication but cannot modify any messages. In this case, Eve gains the same meta information as Alice: Bobs mobile number and its association with Bobs email address. All other information is encrypted on both the email and the SMS channel and so out of reach for Eve.

### Dave the disruptor of service
Dave has access to Alices and Bobs email addresses, Alices and Bobs public keys, and Alices mobile phone number, but he can not read or modify any communication, and has no access to Alices or Bobs private keys. Dave tries to disrupt either the verification process or the encrypted communication between Alice and Bob by either sending a fake key verification SMS to Alice, or a fake key verification response to Bob.

Dave will have no luck, though, because he cannot sucessfully fake either Alices or Bobs signature. And since failed verification attempts simply leave the key in the default "unverified" state, nothing bad happens.

Note: If Alice or Bob would not ignore the failed attempts but interpret it as the detection of a man in the middle and subsequently mark their respective keys as compromised, Dave could sucessfully disrupt encrypted communication between Alice and Bob.

### Mallory the malicious email provider
Mallory can both read and modify all emails between Alice and Bob, and he has access to all public contact information: Alices and Bobs email addresses and public keys, and Alices mobile phone number. Mallory has no access to either the SMS channel nor to Alices or Bobs private keys. 

Mallory executes a classical active man-in-the-middle attack: He pretends to be Bob to Alice and tricks Alice into accepting a public key M<sub>Bob</sub> as Bobs key, and pretends to be Alice to Bob and tricks him to accept a faked M<sub>Alice</sub> key als Alices key.

In this case, the real key verification process will always fail: Bob sends the key verification SMS encrypted with M<sub>Alice</sub>, which Alice will subsequently fail to decrypt.

Of course, Mallory will try to fake the verification process by sending a his own key verification SMS to Alice using M<sub>Bob</sub>, intercept the key verification response email from Alice, and substitute M<sub>Alice</sub> there for Alices key. 

This attempt will fail at Bobs end, though, since it is vanishingly unlikely that Mallory picks the same random number for the faked key verification SMS as Bob chose in the original key verification SMS.

Note: If Alice were to mark Bobs key as "verified" upon receiving a valid key verification SMS, Mallory could trick Alice into accepting M<sub>Bob</sub> as Bobs verified public key.

The consistent failure of the verification process in this scenario is the intended behavior: When key verification always fails in the presence of Mallory, a successful key verification is proof that there is no Mallory in the middle.

### Sally the spy with an IMSI catcher
Sally can access and modify both email and SMS communication between Alice and Bob.

Unfortunately, this means that the key verification process breaks: Sally is able to execute a successful MITM attack on both the email and the SMS channel, and so can easily fake a "successful" key verification by substituting her own keys on both channels.

## Technical considerations

A fully automated key verification process as outlined above requires the ability both send and receive SMS messages from a program. This is possible on Android, but it may not be possible on all relevant OSes. In particular, as far as I know iOS does _not_ allow an app to read SMS contents (it does allow to send an SMS, though). In this case, it is possible to gracefully fall back to a semi-automatic process by including a short notification along the lines of "Paste this message into your Deltachat app" in the cleartext part of the key verification SMS, and including the appropriate UI in Deltachat.

The key verification message will most probably be longer than the 140 ASCII characters allowed in a single SMS message. This should not be a problem, though, since SMS messages can be "chained", and as far as I know all modern smartphone OSs can automatically split a longer text into a series of chained SMS messages on the sending end, and reassemble the full message at the receiving end.

**Comment by hpk42** [comment]
Interesting idea, and considerately written.  Certainly warrants more scrutiny. 

Maybe mentioning e-mail addresses in the content part of the SMS can be avoided?  it'd be nice if the mobile network provider (or anyone who can access the sms content) can not discover e-mail/mobile number relations easily.

**Comment by rkunze** [comment]
Yes, mentioning the e-mail address in clear text can be avoided - actually, I'm not sure what I've been thinking when I put it in the clear text header. Probably something along the lines of giving Alice an incentive to not delete the SMS out of hand as some weird spam in the manual fallback case. But then, in most cases Alice will already have Bobs mobile number and so the SMS will be associated with the contact info for Bob at Alices end even without having the e-mail in clear text.

I've updated the original comment to put Bobs e-mail into the encrypted part of the SMS.
